File: android/LibsqlPlugin.kt
package app.tauri

import android.app.Activity
import android.webkit.WebView
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import app.tauri.plugin.Invoke
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import java.util.UUID
import kotlin.coroutines.CoroutineContext

// 参数类定义
@InvokeArg
internal class PingRequest {
    var value: String? = null
}

@InvokeArg
internal class ConnectOptions {
    lateinit var url: String
    var authToken: String? = null
    var localPath: String? = null
}

@InvokeArg
internal class ExecuteOptions {
    lateinit var connectionId: String
    lateinit var sql: String
    var params: List<Value>? = null
}

@InvokeArg
internal class QueryOptions {
    lateinit var connectionId: String
    lateinit var sql: String
    var params: List<Value>? = null
}

@InvokeArg
internal class SyncOptions {
    lateinit var connectionId: String
}

@InvokeArg
internal class CloseOptions {
    lateinit var connectionId: String
}

// Value 类，用于参数传递
enum class ValueType {
    Null, Integer, Real, Text, Blob
}

@InvokeArg
internal class Value {
    lateinit var type: String
    var value: Any? = null
}

// 数据库连接管理
@TauriPlugin
class LibsqlPlugin(private val activity: Activity) : Plugin(activity), CoroutineScope {
    // 使用 Dispatchers.IO 来执行数据库操作
    override val coroutineContext: CoroutineContext
        get() = Dispatchers.IO

    // 存储数据库连接
    private val connections = mutableMapOf<String, Any>()
    private val databases = mutableMapOf<String, Any>()

    @Command
    fun ping(invoke: Invoke) {
        val request = invoke.parseArgs(PingRequest::class.java)
        val response = JSObject()
        response.put("value", request.value)
        invoke.resolve(response)
    }

    @Command
    fun connect(invoke: Invoke) {
        val options = invoke.parseArgs(ConnectOptions::class.java)
        
        launch {
            try {
                // 这里应该使用 libSQL Kotlin SDK 创建连接
                // 由于我们没有实际的 SDK 引用，这里只是模拟实现
                
                // 创建唯一的连接 ID
                val connectionId = UUID.randomUUID().toString()
                
                // 根据选项创建数据库连接
                // 实际实现会用 libSQL 的 SDK
                // 例如: val db = LibSQL.createClient(options.url, options.authToken)
                
                // 存储连接
                connections[connectionId] = Any() // 模拟数据库连接
                databases[connectionId] = Any()   // 模拟数据库实例
                
                // 返回连接 ID
                val response = JSObject()
                response.put("0", connectionId)
                invoke.resolve(response)
            } catch (e: Exception) {
                invoke.reject(e.message ?: "Failed to connect to database", e.toString())
            }
        }
    }

    @Command
    fun execute(invoke: Invoke) {
        val options = invoke.parseArgs(ExecuteOptions::class.java)
        
        launch {
            try {
                // 检查连接是否存在
                if (!connections.containsKey(options.connectionId)) {
                    invoke.reject("Connection not found", "DATABASE_CONNECTION_NOT_FOUND")
                    return@launch
                }
                
                // 获取连接实例
                // val connection = connections[options.connectionId] as LibSQLConnection
                
                // 执行 SQL 语句
                // 将 options.params 转换为 libSQL 参数格式
                // val result = connection.execute(options.sql, params)
                
                // 模拟执行结果
                val result = JSObject()
                result.put("rowsAffected", 1)
                result.put("lastInsertRowid", 1)
                
                invoke.resolve(result)
            } catch (e: Exception) {
                invoke.reject(e.message ?: "Failed to execute SQL", e.toString())
            }
        }
    }

    @Command
    fun query(invoke: Invoke) {
        val options = invoke.parseArgs(QueryOptions::class.java)
        
        launch {
            try {
                // 检查连接是否存在
                if (!connections.containsKey(options.connectionId)) {
                    invoke.reject("Connection not found", "DATABASE_CONNECTION_NOT_FOUND")
                    return@launch
                }
                
                // 获取连接实例
                // val connection = connections[options.connectionId] as LibSQLConnection
                
                // 执行查询
                // val resultSet = connection.query(options.sql, params)
                
                // 封装结果 (模拟)
                val result = JSObject()
                val columns = JSObject.fromJSONArray("[]")
                val rows = JSObject.fromJSONArray("[]")
                
                result.put("columns", columns)
                result.put("rows", rows)
                
                invoke.resolve(result)
            } catch (e: Exception) {
                invoke.reject(e.message ?: "Failed to execute query", e.toString())
            }
        }
    }

    @Command
    fun sync(invoke: Invoke) {
        val options = invoke.parseArgs(SyncOptions::class.java)
        
        launch {
            try {
                // 检查连接是否存在
                if (!connections.containsKey(options.connectionId)) {
                    invoke.reject("Connection not found", "DATABASE_CONNECTION_NOT_FOUND")
                    return@launch
                }
                
                // 获取数据库实例
                // val database = databases[options.connectionId] as LibSQLDatabase
                
                // 同步远程更改
                // database.sync()
                
                invoke.resolve()
            } catch (e: Exception) {
                invoke.reject(e.message ?: "Failed to sync database", e.toString())
            }
        }
    }

    @Command
    fun close(invoke: Invoke) {
        val options = invoke.parseArgs(CloseOptions::class.java)
        
        launch {
            try {
                // 移除连接
                connections.remove(options.connectionId)
                databases.remove(options.connectionId)
                
                invoke.resolve()
            } catch (e: Exception) {
                invoke.reject(e.message ?: "Failed to close connection", e.toString())
            }
        }
    }
} 

File: ios/LibsqlPlugin.swift
import Foundation
import Tauri
import WebKit

// Request models
struct PingRequest: Decodable {
    let value: String?
}

struct ConnectOptions: Decodable {
    let url: String
    let authToken: String?
    let localPath: String?
}

struct ExecuteOptions: Decodable {
    let connectionId: String
    let sql: String
    let params: [Value]?
}

struct QueryOptions: Decodable {
    let connectionId: String
    let sql: String
    let params: [Value]?
}

struct SyncOptions: Decodable {
    let connectionId: String
}

struct CloseOptions: Decodable {
    let connectionId: String
}

// Value type for parameters
enum ValueType: String, Decodable {
    case Null
    case Integer
    case Real
    case Text
    case Blob
}

struct Value: Decodable {
    let type: String
    let value: AnyCodable?
}

// AnyCodable helper for handling dynamic JSON values
struct AnyCodable: Codable {
    let value: Any
    
    init(value: Any) {
        self.value = value
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if container.decodeNil() {
            self.value = NSNull()
        } else if let bool = try? container.decode(Bool.self) {
            self.value = bool
        } else if let int = try? container.decode(Int.self) {
            self.value = int
        } else if let double = try? container.decode(Double.self) {
            self.value = double
        } else if let string = try? container.decode(String.self) {
            self.value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            self.value = array.map { $0.value }
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            self.value = dict.mapValues { $0.value }
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Cannot decode value")
        }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        switch self.value {
        case is NSNull:
            try container.encodeNil()
        case let bool as Bool:
            try container.encode(bool)
        case let int as Int:
            try container.encode(int)
        case let double as Double:
            try container.encode(double)
        case let string as String:
            try container.encode(string)
        case let array as [Any]:
            try container.encode(array.map { AnyCodable(value: $0) })
        case let dict as [String: Any]:
            try container.encode(dict.mapValues { AnyCodable(value: $0) })
        default:
            throw EncodingError.invalidValue(self.value, EncodingError.Context(codingPath: container.codingPath, debugDescription: "Cannot encode value"))
        }
    }
}

class LibsqlPlugin: Plugin {
    // 存储数据库连接和实例
    private var connections: [String: Any] = [:]
    private var databases: [String: Any] = [:]
    
    @objc public func ping(_ invoke: Invoke) throws {
        let request = try invoke.parseArgs(PingRequest.self)
        invoke.resolve(["value": request.value as Any])
    }
    
    @objc public func connect(_ invoke: Invoke) throws {
        let options = try invoke.parseArgs(ConnectOptions.self)
        
        // 在实际实现中，这里应该使用 libSQL Swift 客户端
        // 为简化示例，我们仅模拟实现
        
        // 创建连接 ID
        let connectionId = UUID().uuidString
        
        // 创建一个 DispatchQueue 在后台线程操作数据库
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            // 模拟创建数据库连接
            // 实际实现应该使用 LibSQL 客户端
            // let db = LibSQL.connect(url: options.url, authToken: options.authToken, localPath: options.localPath)
            
            // 存储连接和数据库实例
            self.connections[connectionId] = "connection"  // 模拟连接
            self.databases[connectionId] = "database"      // 模拟数据库
            
            // 返回连接 ID
            DispatchQueue.main.async {
                invoke.resolve([connectionId])
            }
        }
    }
    
    @objc public func execute(_ invoke: Invoke) throws {
        let options = try invoke.parseArgs(ExecuteOptions.self)
        
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            // 检查连接是否存在
            guard self.connections[options.connectionId] != nil else {
                DispatchQueue.main.async {
                    invoke.reject("Connection not found", "DATABASE_CONNECTION_NOT_FOUND")
                }
                return
            }
            
            // 在实际实现中，获取连接并执行 SQL
            // let connection = self.connections[options.connectionId] as! LibSQLConnection
            // let params = self.convertParams(options.params)
            // let result = try? connection.execute(options.sql, params: params)
            
            // 模拟执行结果
            DispatchQueue.main.async {
                invoke.resolve([
                    "rowsAffected": 1,
                    "lastInsertRowid": 1
                ])
            }
        }
    }
    
    @objc public func query(_ invoke: Invoke) throws {
        let options = try invoke.parseArgs(QueryOptions.self)
        
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            // 检查连接是否存在
            guard self.connections[options.connectionId] != nil else {
                DispatchQueue.main.async {
                    invoke.reject("Connection not found", "DATABASE_CONNECTION_NOT_FOUND")
                }
                return
            }
            
            // 在实际实现中，获取连接并执行查询
            // let connection = self.connections[options.connectionId] as! LibSQLConnection
            // let params = self.convertParams(options.params)
            // let rows = try? connection.query(options.sql, params: params)
            
            // 模拟查询结果
            DispatchQueue.main.async {
                invoke.resolve([
                    "columns": [],
                    "rows": []
                ])
            }
        }
    }
    
    @objc public func sync(_ invoke: Invoke) throws {
        let options = try invoke.parseArgs(SyncOptions.self)
        
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            // 检查数据库是否存在
            guard self.databases[options.connectionId] != nil else {
                DispatchQueue.main.async {
                    invoke.reject("Database not found", "DATABASE_NOT_FOUND")
                }
                return
            }
            
            // 在实际实现中，获取数据库并同步
            // let database = self.databases[options.connectionId] as! LibSQLDatabase
            // try? database.sync()
            
            DispatchQueue.main.async {
                invoke.resolve(nil)
            }
        }
    }
    
    @objc public func close(_ invoke: Invoke) throws {
        let options = try invoke.parseArgs(CloseOptions.self)
        
        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            guard let self = self else { return }
            
            // 移除连接和数据库
            self.connections.removeValue(forKey: options.connectionId)
            self.databases.removeValue(forKey: options.connectionId)
            
            DispatchQueue.main.async {
                invoke.resolve(nil)
            }
        }
    }
    
    // 助手方法：转换参数
    private func convertParams(_ params: [Value]?) -> [Any] {
        guard let params = params else { return [] }
        
        return params.map { param in
            switch param.type {
            case "Null":
                return NSNull()
            case "Integer":
                if let value = param.value?.value as? Int {
                    return value
                }
                return 0
            case "Real":
                if let value = param.value?.value as? Double {
                    return value
                }
                return 0.0
            case "Text":
                if let value = param.value?.value as? String {
                    return value
                }
                return ""
            case "Blob":
                if let value = param.value?.value as? [UInt8] {
                    return Data(value)
                }
                return Data()
            default:
                return NSNull()
            }
        }
    }
} 